config/base/derive/src/proxy.rs

```diff
@@ -140,30 +116,76 @@ fn impl_load_from_disk(proxy_name: &Ident) -> proc_macro2::TokenStream {
     }
 }
 
-fn gen_proxy_struct_fields(ast: &StructWithFields) -> proc_macro2::TokenStream {
-    let combined_fields = ast.fields.iter().map(|field| {
-        let ident = &field.ident;
-        let ty = &field.ty;
-        quote! {
-            #ident: Option<#ty>
+fn gen_proxy_struct(mut ast: StructWithFields) -> StructWithFields {
+    // Make mut, parse quote and modify ast for proxy
+    ast.fields.iter_mut().for_each(|field| {
+        // For fields of `Configuration` that have an inner config, the corresponding
+        // proxy field should have a proxy there as well
+        if field.has_inner {
+            #[allow(clippy::expect_used)]
+            if let Type::Path(path) = &mut field.ty {
+                let old_ident = &path.path.segments.last().expect("Can't be empty").ident;
+                let new_ident = format_ident!("{}Proxy", old_ident);
+                path.path.segments.last_mut().expect("Can't be empty").ident = new_ident;
+            }
         }
+        let ty = &field.ty;
+        field.ty = parse_quote! {
+            Option<#ty>
+        };
+        field.has_option = true;
+        // Also remove `#[serde(default = ..)]` if present
+        // as it breaks proxy deserialization
+        utils::remove_attr(&mut field.attrs, "serde");
     });
+    ast.parent_ident = Some(ast.ident.clone());
+    ast.ident = format_ident!("{}Proxy", ast.ident);
+    utils::remove_attr_struct(&mut ast, "config");
+    ast
+}
+
+// REVIEW: need lvalue here as well?
```

Comment:
Could you explain what you mean here?

Reply:
The `config_derive` crate still has logic related to types inside `Arc<RwLock>`s, such as calling `read()` and awaiting or calling `write()` on the field. It was there initially and I only refactored it away into `utils` mod. Even though not a single component is wrapped inside that in our codebase now, @appetrosyan mentioned it might be prudent to still have it -- for better generality, I suppose. And these `lvalue`s are currently used in e.g. `LoadFromEnv` impl. Thus what I mean here is that probably `gen_none_fields_check` needs to incorporate that as well.

Reply:
Ok, i get it.
Our proxy configs are used only for obtaining original configs from different sources as i understand, so it's hard to imagine that their fields would be wrapped into something like `Arc<RwLock>`.